### ISSUE TITLE
reduce size of response cache for typical workloads

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#626]

**Comment by tschottdorf** [comment]
the response cache currently caches all mutating requests for a long time and accrues a significant amount of data which almost certainly isn't ever accessed by the client.
our goals here are

- to offer replay protection by default (no opt-in, which invites misconfiguration)
- quickly expunge redundant information (not true so far).

the most likely useful method for this is putting into place a system by which clients acknowledge commands which they have executed successfully. if we key response cache entries (partly) by client ID and a client-supplied seq, an ideal system could expunge all entries prior to the current sequence number. due to parallel requests sent by the client, this isn't going to be as easy.

another related circle of ideas is to not have the response cache on each range individually, but on each store. with some implementational changes, this would obviate the need to copy portions of the cache on each range split, and speed up the `splitTrigger` significantly.

original title/suggestion of this issue:

* Response cache should use 1 hour since last committed log entry for GC

**Comment by bdarnell** [comment]
One more possibility that we've discussed: have the client send an explicit ack for each ClientCmdID it has sent (possibly at the end of the transaction, or after each individual command). This probably wouldn't be good for non-multiplexed transports like HTTP/1. It's also tricky because the client would need to be given the range ID that handled each request (or re-send the keys - maybe this could be bundled with intent resolution, although we were planning to do intent resolution on the coordinator)

**Comment by tschottdorf** [comment]
I think a good start for this is implementing the strategy in [this post](https://github.com/cockroachdb/cockroach/issues/2297#issuecomment-147717034/), but still keyed under `ReplicaID`: `/prefix/<replica_id>/<cmd_id>/<{start,end}_key>` -> `BatchResponse, error`. Then we avoid inflating the size of the response cache on a split: instead of copying the whole response cache into the new range, we move those responses belonging to the new range (keeping range-spanning ones in the old one) over. On a merge, we only move the subsumed replica's cache into that of the subsuming one (we'll combine responses on the cache hit). As a side effect, that also deals with #2297.

Let me know if that sounds good and I'll get started.

To actually minimize the size of the cache there are some options (none of which I'm immediately beginning work on, so discussing them has a bit more time):


* The response cache is only "really" interesting for transactions (where writes see the latest intent, and so both Raft and clients need replay protection). That is also going to be the most common use case going forward (each SQL statement goes into a Transaction). Since the coordinator already collects the intents, it can also collect the associated `CmdID`s and send them along with `EndTransaction`, and finally (via `Replica`) the resulting `ResolveIntent{,Range}` calls, which as a side effect of clearing the intent would GC the corresponding response cache entries. The caveat here is that we still might want to cache the batch which has the `EndTransaction` call to get replay protection on the commit, or the following could happen: client runs a batch which includes the final `EndTransaction`; it commits but the client never gets the reply. It retries, finding that it's already been committed - but is unable to access the responses from the original batch (only relevant if there's a read). It's not clear how the SQL gateway would handle this - it can't re-run the thing (since it already ran) but it also can't report anything back to the client. I seriously doubt that Postgres and friends have any solution for that - maybe it's acceptable to return an error every once in a blue moon.

* Raft may commit proposed commands multiple times. In that case, we don't actually have a client waiting for that command anywhere, and so there's no reason to cache the response unless re-executing the request changes the state of the database. Any non-transactional write (weirder commands ignored for now) which succeeds will fail the second time around (WriteTooOldError), so a batch only needs to be cached if it doesn't contain such an operation. Commands which error out need a little more love: a `CPut` which fails at first could succeed on the second try (if the underlying value has changed), I think, so probably we would want to cache (some of?) those just to be safe.
In particular, removing replay protection for `ResolveIntent{Range,}`, `GCRequest`, `TruncateLog` makes sense (and the first one should be a relevant improvement).

**Comment by bdarnell** [comment]
> I think a good start for this is implementing the strategy in this post, but still keyed under `ReplicaID`

You mean `RangeID`, right? 

What does `<{start,end}_key>` mean? Is it `<start_key>/<end_key>` or are the keys combined in some other way? Is the end key really needed here? I think we just need the start key for the full response cache entries. The "sentinel" entries for range requests need both the start and end keys to correctly propagate on splits, but that would go in the value rather than the key.

The first paragraph LGTM; I'll come back to the rest later.

**Comment by bdarnell** [comment]
> it commits but the client never gets the reply

I believe this scenario is common to nearly all databases, and the only workaround I know of is to be less aggressive about batching: as long as the final `COMMIT` is sent by itself (which is the common case for many data access layers) you can ensure that you have all of the individual statement responses before committing the transaction. 

> a CPut which fails at first could succeed on the second try (if the underlying value has changed)

But wouldn't it fail with a WriteTooOldError just like a non-conditional put? 

One concern with this treatment of non-transactional writes is that we currently allow commands without timestamps. The server will assign a timestamp which should never get a `WriteTooOldError`. But now that the KV API is internal-only we could probably just make the timestamp required and never fill it in at the `Store`.

**Comment by tschottdorf** [comment]
yes, sorry, `RangeID`. What I meant with the `{start,end}` above is actually two keys:

```
/prefix/<replica_id>/<cmd_id>/<start_key> -> actual response
/prefix/<replica_id>/<cmd_id>/<end_key> -> nil (sentinel)
```

The reason I need this is that when the range splits, the second range still needs to be able to figure out that it's supposed to return an empty reply, not execute a partial request. That isn't enough (split the range again, and you have the two keys above outside of the middle range if you're unlucky), so additionally, on a split, whenever a sentinel "slips" into the right range, we write a sentinel at <desc_end_key> of the left range. Merges, on the other hand, can drop any "intermediate" sentinels, but that doesn't matter in practice.

Demonstration: range request on `[c,f)` gets cached on range `[a,z)`:

```
/id/c -> resp
/id/f -> original sentinel
```

Range splits at `d`:

```
# [a,d)
/id/c -> resp
/id/d -> sentinel // this one technically doesn't need to be written
# [d,z)
/c/f -> sentinel
```

`[d,z)` splits at `e`:

```
# [a,d)
/id/c -> resp
/id/d -> sentinel
# [d,e)
/c/e -> sentinel
# [e, z)
/c/f -> sentinel
```

I think that should work, though I'm going to sleep on it to check whether it still sounds reasonable tomorrow. It's a shame we have to do so much work here. Feedback appreciated.

**Comment by tschottdorf** [comment]
> But wouldn't it fail with a WriteTooOldError just like a non-conditional put?

Say at `t`, a `CPut` executes and fails (whatever is present doesn't match). Now if something else gets written at a timestamp `s < t`, then we're doomed because the `CPut` could come back (still at `t`) and suddenly find something it likes. So the question is really whether a failed `CPut` prevents a write to go "under" the failed attempt, and I don't think so because the timestamp cache only updates on successful operations.

> One concern with this treatment of non-transactional writes is that we currently allow commands without timestamps. The server will assign a timestamp which should never get a WriteTooOldError. But now that the KV API is internal-only we could probably just make the timestamp required and never fill it in at the Store.

Internally, we can (and I think already do) set timestamps fairly early (maybe not early enough), though that's something we'd want to double check. At the latest, the store assigns a timestamp to the best of my knowledge (I changed a bunch of things there during the batch refactors).

**Comment by bdarnell** [comment]
Also from our conversation last night: if we don't care about replay protection for non-transactional commands, we may be able to get rid of the response cache completely. The idea is that transactional writers must be prepared to retry their transactions in any case, so we just need to be able to fail replayed requests instead of recovering their original response. The half-baked idea is to use the write intent as the response cache:
* Command IDs are no longer global, they're scoped to the transaction. Use a monotonic sequence assigned by the coordinator
* Store the command ID along with the transaction ID in the write intent.
* Fail any command which would replace an intent from a newer command with an older one.

This would greatly simplify things, at the expense of converting some single-command retries to full-transaction retries. (and possibly other downsides we haven't thought through yet)

**Comment by tschottdorf** [comment]
yes, I liked that idea. Even in the case in which we can't completely get rid of non-txn response cache, we could still explore that further. The great thing in that setting is that a guarantee that an error is returned for the aggregate request allows us to not worry about executing individual bits of it (we'll leave dangling intents which aren't tracked though, that's something on my radar for the eventual `DistSender` refactor).

**Comment by tschottdorf** [comment]
Small addenda: we discussed in person that the `CPut` anomaly is really due to the fact that a `CPut` which "fails" on the grounds that the read value didn't match doesn't really constitute an error: it's rather a positive result with the information that the previous value does not match and no write was performed. If that were the case, the read timestamp cache would update and the anomaly described above vanish.

In short, there should be no `ConditionFailedError`. Unfortunately, that's a sizeable refactor since we rely on current `CPut` behaviour throughout the SQL layer. We can make ends meet for now by special-casing that error when updating the timestamp cache, but should clean this up in the long run. But for the purposes of this issue, let's assume that we're special casing the error.

**Comment by tschottdorf** [comment]
done with #3157, #3126, #3144, #3140.
